File: BETELITE_v3/BETELITE_v3/backend/routes/lobby.js
const express = require('express');
const { v4: uuid } = require('uuid');

// In-memory lobby challenges
// Map of challengeId -> challenge details
const lobbyChallenges = new Map();

module.exports = (io, engine, db) => {
  const router = express.Router();

  // Get all active open challenges
  router.get('/', (req, res) => {
    try {
      const challenges = Array.from(lobbyChallenges.values());
      res.json({ ok: true, challenges });
    } catch (e) {
      res.status(500).json({ error: e.message });
    }
  });

  // Create a new challenge
  router.post('/create', async (req, res) => {
    try {
      const { uid, username, game, amount, platform } = req.body;
      if (!uid || !amount || !game) return res.status(400).json({ error: 'Missing fields' });

      // Check creator balance before allowing challenge post
      if (db) {
        const snap = await db.ref(`users/${uid}/wallet/balance`).once('value');
        const balance = snap.val() || 0;
        if (balance < Number(amount)) {
          return res.status(400).json({ error: 'Insufficient balance to post this challenge' });
        }
      }

      const challengeId = `ch_${uuid().substring(0, 8)}`;
      const challenge = {
        id: challengeId,
        creatorId: uid,
        creatorName: username || 'Unknown',
        game,
        amount: Number(amount),
        platform: platform || 'mobile',
        createdAt: Date.now()
      };

      lobbyChallenges.set(challengeId, challenge);
      
      // Broadcast to lobby
      io.emit('lobby_new_challenge', challenge);

      res.json({ ok: true, challenge });
    } catch (e) {
      res.status(500).json({ error: e.message });
    }
  });

  // Delete a challenge (only by creator)
  router.post('/delete', (req, res) => {
    try {
      const { challengeId, uid } = req.body;
      const challenge = lobbyChallenges.get(challengeId);
      if (!challenge) return res.status(404).json({ ok: false, error: 'Wager not found' });
      if (challenge.creatorId !== uid) return res.status(403).json({ ok: false, error: 'You can only delete your own wagers' });
      
      lobbyChallenges.delete(challengeId);
      io.emit('lobby_challenge_removed', { id: challengeId });
      console.log(`[LOBBY] Wager ${challengeId} deleted by ${uid}`);
      res.json({ ok: true });
    } catch (e) {
      res.status(500).json({ ok: false, error: e.message });
    }
  });
  // Accept a challenge
  router.post('/accept', async (req, res) => {
    try {
      const { challengeId, uid, username } = req.body;
      const challenge = lobbyChallenges.get(challengeId);
      
      if (!challenge) return res.status(404).json({ error: 'Challenge not found or already accepted' });
      if (challenge.creatorId === uid) return res.status(400).json({ error: 'Cannot accept your own challenge' });

      // Atomically deduct from both wallets into escrow using transactions
      if (db) {
        // Validate and deduct creator balance atomically
        const creatorBalRef = db.ref(`users/${challenge.creatorId}/wallet/balance`);
        const creatorResult = await creatorBalRef.transaction((currentBalance) => {
          const bal = currentBalance || 0;
          if (bal < challenge.amount) return; // Abort transaction (returns undefined)
          return bal - challenge.amount;
        });
        if (!creatorResult.committed) {
          lobbyChallenges.delete(challengeId);
          io.emit('lobby_challenge_removed', { id: challengeId });
          return res.status(400).json({ error: 'Challenger no longer has sufficient funds' });
        }

        // Validate and deduct acceptor balance atomically
        const acceptorBalRef = db.ref(`users/${uid}/wallet/balance`);
        const acceptorResult = await acceptorBalRef.transaction((currentBalance) => {
          const bal = currentBalance || 0;
          if (bal < challenge.amount) return; // Abort
          return bal - challenge.amount;
        });
        if (!acceptorResult.committed) {
          // Refund creator since we already deducted
          await creatorBalRef.transaction((currentBalance) => {
            return (currentBalance || 0) + challenge.amount;
          });
          return res.status(400).json({ error: 'Insufficient balance to accept this challenge' });
        }

        // Record escrow
        const escrowId = `esc_${challengeId}`;
        await db.ref(`escrow/${escrowId}`).set({
          challengeId,
          creatorId:  challenge.creatorId,
          acceptorId: uid,
          amount:     challenge.amount,
          pool:       challenge.amount * 2,
          status:     'held',
          heldAt:     Date.now(),
        });
        console.log(`[ESCROW] ₦${challenge.amount * 2} held for challenge ${challengeId}`);

        // Log transaction for audit trail
        await db.ref('transactions').push().set({
          type:         'escrow_hold',
          amount:       challenge.amount * 2,
          participants: [challenge.creatorId, uid],
          escrowId,
          challengeId,
          timestamp:    Date.now(),
        });
      }

      // Create Match in GameEngine
      const match = engine.createP2PMatch({
        gameType: challenge.game,
        home: challenge.creatorName,
        away: username || 'Opponent',
        homeId: challenge.creatorId,
        awayId: uid,
        wagerPool: challenge.amount * 2,
        wagerAmount: challenge.amount,
        status: 'live' // auto start
      });

      // Remove from lobby
      lobbyChallenges.delete(challengeId);
      
      // Broadcast match created so both clients can redirect
      io.emit('lobby_challenge_accepted', {
        challengeId,
        matchId: match.id,
        creatorId: challenge.creatorId,
        acceptorId: uid
      });

      res.json({ ok: true, matchId: match.id });
    } catch (e) {
      res.status(500).json({ error: e.message });
    }
  });

  return router;
};

File: BETELITE_v3/BETELITE_v3/backend/routes/lobby.go
package routes

import (
	"context"
	"fmt"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/livekit/protocol/auth"

	"betelite-go/config"
	"betelite-go/db"
	"betelite-go/middleware"
	"betelite-go/models"
	"betelite-go/services"
	"betelite-go/utils"
	"betelite-go/ws"
)

func SetupLobbyRoutes(api fiber.Router, hub *ws.Hub) {
	lobby := api.Group("/lobby", middleware.AuthRequired())

	// Create a new challenge (Escrow hold)
	lobby.Post("/challenge", func(c *fiber.Ctx) error {
		var req struct {
			Game     string `json:"game"`
			Amount   int64  `json:"amount"` // Note: Frontend might send string/float, ensure it's converted to int64 kobo
			Username string `json:"username"`
		}
		if err := c.BodyParser(&req); err != nil {
			return utils.SendError(c, 400, "Invalid payload")
		}

		uid := middleware.GetUID(c)
		challengeID := fmt.Sprintf("challenge_%d", time.Now().UnixNano())

		// DB Transaction: Deduct wallet and create escrow
		ctx := context.Background()
		tx, err := db.Pool.Begin(ctx)
		if err != nil {
			return utils.SendError(c, 500, "Database error")
		}
		defer tx.Rollback(ctx)

		// Check balance
		var balance int64
		err = tx.QueryRow(ctx, "SELECT balance FROM users WHERE id = $1 FOR UPDATE", uid).Scan(&balance)
		if err != nil || balance < req.Amount {
			return utils.SendError(c, 400, "Insufficient funds")
		}

		err = services.AdjustBalance(ctx, tx, uid, -req.Amount, "wager_hold", challengeID)
		if err != nil {
			return utils.SendError(c, 500, "Failed to update balance")
		}

		// Create Escrow
		_, err = tx.Exec(ctx, `INSERT INTO escrow (challenge_id, creator_id, amount, pool, status) 
			VALUES ($1, $2, $3, $4, 'waiting')`,
			challengeID, uid, req.Amount, req.Amount) // Pool is just creator's amount for now
		if err != nil {
			return utils.SendError(c, 500, "Failed to create escrow")
		}

		if err := tx.Commit(ctx); err != nil {
			return utils.SendError(c, 500, "Transaction commit failed")
		}

		challenge := models.Challenge{
			ID:          challengeID,
			CreatorID:   uid,
			CreatorName: req.Username,
			Game:        req.Game,
			Amount:      req.Amount,
			Currency:    "NGN",
			Timestamp:   time.Now().UnixMilli(),
		}

		// Broadcast new challenge
		ws.BroadcastEvent(hub, "lobby_update", challenge)

		return utils.SendSuccess(c, fiber.Map{
			"challengeId": challengeID,
		})
	})

	// Accept challenge
	lobby.Post("/challenge/accept", func(c *fiber.Ctx) error {
		var req struct {
			ChallengeID string `json:"challengeId"`
			Username    string `json:"username"`
		}
		if err := c.BodyParser(&req); err != nil {
			return utils.SendError(c, 400, "Invalid payload")
		}

		uid := middleware.GetUID(c)
		ctx := context.Background()

		tx, err := db.Pool.Begin(ctx)
		if err != nil {
			return utils.SendError(c, 500, "Database error")
		}
		defer tx.Rollback(ctx)

		// Get Escrow details
		var escrow models.Escrow
		var status string
		var player1ID string
		var amount int64
		err = tx.QueryRow(ctx, "SELECT id, creator_id, amount, status FROM escrow WHERE challenge_id = $1 FOR UPDATE", req.ChallengeID).
			Scan(&escrow.ID, &player1ID, &amount, &status)
		if err != nil || status != "pending" {
			return utils.SendError(c, 400, "Challenge not available")
		}

		if player1ID == uid {
			return utils.SendError(c, 400, "Cannot accept your own challenge")
		}

		// Check acceptor's balance
		var balance int64
		err = tx.QueryRow(ctx, "SELECT balance FROM users WHERE id = $1 FOR UPDATE", uid).Scan(&balance)
		if err != nil || balance < amount {
			return utils.SendError(c, 400, "Insufficient funds")
		}

		err = services.AdjustBalance(ctx, tx, uid, -amount, "wager_hold", req.ChallengeID)
		if err != nil {
			return utils.SendError(c, 500, "Failed to update balance")
		}

		// Update Escrow
		matchID := fmt.Sprintf("match_%d", time.Now().UnixNano())
		pool := amount * 2
		_, err = tx.Exec(ctx, "UPDATE escrow SET acceptor_id = $1, pool = $2, status = 'held', match_id = $3 WHERE id = $4",
			uid, pool, matchID, escrow.ID)
		if err != nil {
			return utils.SendError(c, 500, "Failed to update escrow")
		}

		if err := tx.Commit(ctx); err != nil {
			return utils.SendError(c, 500, "Transaction commit failed")
		}

		// Create match in Engine
		match := &models.Match{
			ID:           matchID,
			Game:         "EA FC 24", // Would normally pull from challenge
			HomeID:       escrow.CreatorID,
			AwayID:       uid,
			Status:       "live",
			Minute:       0,
			WagerPool:    pool,
			WagerAmount:  escrow.Amount,
			IsP2P:        true,
			ChallengeID:  req.ChallengeID,
		}
		services.Engine.AddMatch(match)

		// Broadcast removal of challenge
		ws.BroadcastEvent(hub, "lobby_challenge_removed", map[string]string{"id": req.ChallengeID})

		return utils.SendSuccess(c, fiber.Map{"matchId": matchID})
	})

	// LiveKit Token Generation
	lobby.Post("/stream/token", func(c *fiber.Ctx) error {
		var req struct {
			RoomName string `json:"roomName"`
			Identity string `json:"identity"`
			IsHost   bool   `json:"isHost"`
		}
		if err := c.BodyParser(&req); err != nil {
			return utils.SendError(c, 400, "Invalid payload")
		}

		apiKey := config.Cfg.LiveKitAPIKey
		apiSecret := config.Cfg.LiveKitAPISecret
		if apiKey == "" || apiSecret == "" {
			return utils.SendError(c, 500, "LiveKit keys not configured")
		}

		at := auth.NewAccessToken(apiKey, apiSecret)
		grant := &auth.VideoGrant{
			RoomJoin: true,
			Room:     req.RoomName,
		}
		if req.IsHost {
			grant.CanPublish = &[]bool{true}[0]
			grant.CanPublishData = &[]bool{true}[0]
		} else {
			grant.CanPublish = &[]bool{false}[0]
			grant.CanPublishData = &[]bool{true}[0] // allow chatting
		}

		at.AddGrant(grant).
			SetIdentity(req.Identity).
			SetValidFor(time.Hour * 4)

		token, err := at.ToJWT()
		if err != nil {
			return utils.SendError(c, 500, "Failed to create token")
		}

		return utils.SendSuccess(c, fiber.Map{"token": token})
	})
}
